### PR TITLE
search ui: move search history backend to temporary settings

### PR DESCRIPTION
**Pull request by limitedmage**
Search history is now stored in temporary settings, with a max of 20 items in the list. This is an MRU list: any newer item removes the oldest item in the list when at the limit.

Additionally, items are only added to the search history if the user actually executes the search from the search box on the homepage or navbar. Navigating to a search via a link or other means will not add the search to the history.

## Test plan

- Added numerous unit tests for the new recent searches hook and autocompletion.
- Lots of manual testing 😨 There are a lot of edge cases and timing issues that are hard to repro for this feature.

## App preview:

- [Web](https://sg-web-jp-recentsearchestempsettings.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-qjwpbltgwb.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Review by novoselrok — APPROVED** [review]
LGTM, left a couple of minor comments.

File: client/web/src/search/input/useRecentSearches.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+// Returns all recent searches from temporary settings and a function to add a new search to the list.
+// If no recent searches exist, the temporary settings is initialized with
+// the user's recent searches from the event log.
+export function useRecentSearches(): {
+    recentSearches: RecentSearch[] | undefined
+    addRecentSearch: (query: string) => void
+    state: 'loading' | 'error' | 'success'
+} {
+    const [recentSearches, setRecentSearches] = useTemporarySetting('search.input.recentSearches', [])
+    const [state, setState] = useState<'loading' | 'error' | 'success'>('loading')
```
[the remark this answers is not in the document]
Comment (limitedmage) [reply client/web/src/search/input/useRecentSearches.ts:35]: I had originally set the state to "error" but then decided against it to instead fail gracefully. I'll remove the error state from the typing for now and we can add it back if we need to.

File: client/web/src/search/input/useRecentSearches.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    // If the list is full, the oldest search is removed.
+    const addOrMoveRecentSearchToTop = useCallback(
+        (recentSearch: RecentSearch) => {
+            setRecentSearches(recentSearches => {
+                const newRecentSearches = recentSearches?.filter(search => search.query !== recentSearch.query) || []
+                newRecentSearches.unshift(recentSearch)
+                // Truncate array if it's too long
+                if (newRecentSearches.length > MAX_RECENT_SEARCHES) {
+                    newRecentSearches.length = MAX_RECENT_SEARCHES
+                }
```
[the remark this answers is not in the document]
Comment (limitedmage) [reply client/web/src/search/input/useRecentSearches.ts:85]: Yeah it's a weird JS thing 😅 `slice` works too and I agree it's more readable.

File: client/web/src/search/input/useRecentSearches.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                    setState('success')
+                })
+                .catch(() => {
+                    logger.error('Error fetching recent searches from event log')
+                    setState('success') // Ignore the error and use the empty list from temporary settings
+                })
+        }
+    }, [recentSearches, loadFromEventLog, setRecentSearches])
+
+    useEffect(() => {
```
[the remark this answers is not in the document]
Comment (limitedmage) [reply client/web/src/search/input/useRecentSearches.ts]: Thanks, this makes sense!

**Pull request merged**